Userland/Utilities/dd.cpp

```diff
@@ -0,0 +1,118 @@
+#include <AK/Optional.h>
+#include <AK/String.h>
+#include <AK/Vector.h>
+
+#include <fcntl.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <stdio.h>
+#include <string.h>
+
+const char* usage = "usage: dd <options>\n\n"
+                    "options:\n"
+                    "if=<file>\t\tinput file (default: stdin)\n"
+                    "of=<file>\t\toutput file (default: stdout)\n"
+                    "bs=<number>\t\tblocks size (default: 512)\n"
+                    "count=<number>\t\tnumber of blocks to copy\n"
+                    "--help\t\tshows this text\n";
+
+static String split_at_equals(const char* argument)
+{
+    String string_value(argument);
+
+    auto values = string_value.split('=');
+    if (values.size() != 2) {
+        fprintf(stderr, "Unable to parse: %s\n", argument);
+        return {};
+    } else {
+        return values[1];
+    }
+}
+
+static int handle_io_file_arguments(int& fd, const char* argument)
+{
+    auto value = split_at_equals(argument);
+    if (value.is_empty()) {
+        return -1;
+    }
+
+    fd = open(value.characters(), O_RDONLY);
+    if (fd == -1) {
+        fprintf(stderr, "Unable to open: %s\n", value.characters());
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+static int handle_numeric_arguments(int& numeric_value, const char* argument)
+{
+    auto value = split_at_equals(argument);
+    if (value.is_empty()) {
+        return -1;
+    }
+
+    Optional<int> numeric_optional = value.to_int();
+    if (!numeric_optional.has_value()) {
+        fprintf(stderr, "Invalid numeric value: %s\n", value.characters());
+        return -1;
+    }
+
+    numeric_value = numeric_optional.value();
+
+    if (numeric_value <= 0) {
+        fprintf(stderr, "Invalid numerical value: %d\n", numeric_value);
+        return -1;
+    } else {
+        return 0;
+    }
+}
+
+int main(int argc, char** argv)
+{
+    int input_fd = 0;
+    int output_fd = 1;
+    int block_size = 512;
+    int count = -1; // infinite
+
+    // Handling heretic command-line options.
+    for (int a = 1; a < argc; a++) {
```

Comment:
better to use  Core::ArgsParser to parse args.

Reply:
Ohh, I thought that ArgParser only deals with arguments in the format of "-a value" or "--arg value" and dd uses a different format.

Reply:
You are right, I didn't aware of this😄.